### DISCUSSION TITLE
Error only after running `next build`

### DISCUSSION THREAD

**Discussion by kblizeck** [vercel/next.js#63955] [Help] [answered]
### Summary

Encountering a javascript error only in production mode, after running `next build`. In Dev mode, everything works as expected.

[Apryse WebViewer package](https://www.npmjs.com/package/@pdftron/webviewer) (we purchased a license with them).

Importing in a client side component, so had to import in useEffect to avoid `window is not defined` error:

```
useEffect(() => {
    import('@pdftron/webviewer').then(() => {
      WebViewer(
        {
          licenseKey: process.env.NEXT_PUBLIC_PDFTRON_WEBVIEWER_LICENSE_KEY,
          path: '/WebViewer/10.8.0/lib',
          preloadWorker: 'pdf',
          initialDoc: URL_TO_AWS_S3_FILE,
          filename: `${paperTitle}.pdf`,
        },
        viewer.current
      ).then(async (instance) => {
        // This is populated from my database in the app
        const teamUsers = [
          {
            value: 'John Doe',
            email: 'john@johndoe.com',
          },
          {
            value: 'Jane Doe',
            email: 'jane@janedoe.com'
          }
        ];
        instance.UI.mentions.setUserData(teamUsers);
        instance.UI.mentions.setAllowedTrailingCharacters([' ', ',']);

        instance.UI.mentions.on('mentionChanged', (mentions, action) => {
          sendMentionChange(mentions, action);
        });
      });
    });
  }, []);

  return (
    <div
      className="webviewer"
      ref={viewer}
      style={{ height: 'calc(100vh - 75px)', width: '100%' }}
    ></div>
  );
```

Their package uses [`react-quill`](https://www.npmjs.com/package/react-quill) which in turn uses the [`quill-mention`](https://www.npmjs.com/package/quill-mention) module

You can view the WebViewer source code here - https://github.com/PDFTron/webviewer-ui

When loading the Webviewer, you can send in an array of user data that will populate the quill-mention list in react quill. In Nextjs dev mode - everything works as expected.

After a Production build - `next build` - I get a getBoundingClientRect error, traced back to the minifed version of the quill editor:
![Screenshot 2024-03-27 at 12 49 05 PM](https://github.com/vercel/next.js/assets/9886123/83202384-9d3f-4ecf-9e52-710295cb0cb8)

`CommentTextArea` is a script in the WebViewer source code that loads the ReactQuill component, with a quill-mention module extension. Follows the Quill mention documentation exactly.

[CommentTextArea ReactQuill instantiation](https://github.com/PDFTron/webviewer-ui/blob/44babe17df1e65e37a1fd97ee769093b50dbc954/src/components/NoteTextarea/CommentTextarea/CommentTextarea.js#L151)

I'm assuming it's some sort of tree-shaking that's happening during Nextjs production build, since the error is only occurring after a build.

Trying to figure out how to exclude this from the build maybe?

**Comment by icyJoseph** [comment]
One way to debug this is to disable SWC minification, it'll print a warning about SWC minification controls being disabled, but this would be a good indicator that something is going south during minification.

**Comment by kblizeck** [comment]
I tried disabling that already in next.config.js and the error is still
occurring after a build. Still not occurring before the build in dev mode…

**Reply by icyJoseph** [reply]
Yeah but the minification happens only in build time, that's why it is a good check.

**Comment by kblizeck** [comment]
Think I’m chasing the right herring, if disabling minification didn’t fix
the error?

**Reply by icyJoseph** [reply]
I think, it could then be a problem with refs. I seem to remember some bug report about refs having different behavior in prod. Perhaps react strict more is at play during dev mode? 

Also try to debug/log the value of the ref

**Reply by kblizeck** [reply, marked as answer]
@icyJoseph I tracked it back to a type check in `quill.js` - 
<img width="418" alt="Screenshot 2024-04-02 at 8 49 29 PM" src="https://github.com/vercel/next.js/assets/9886123/e74392bd-d811-47b4-a18a-c86dbaf7616c">

The `ref` being sent in was a `nodeType: text`, but this `instanceof` wasn't catching it. I added a `nodeType` check - `if (node instanceof Text || node.nodeType === Node.TEXT_NODE)` - and this solved the error!
